File: library/src/main/java/cn/wandersnail/widget/listener/RejectFastClickListener.java
package cn.wandersnail.widget.listener;

import android.util.SparseLongArray;
import android.view.View;

import androidx.annotation.NonNull;

import java.util.Objects;

/**
 * 防止快速点击的
 * <p>
 * date: 2019/8/22 16:43
 * author: zengfansheng
 */
public abstract class RejectFastClickListener {
    private int minInterval = 1000;
    private SparseLongArray clickTimeArr = new SparseLongArray();

    public RejectFastClickListener(int minInterval) {
        if (minInterval > 0) {
            this.minInterval = minInterval;
        }        
    }

    protected void check(@NonNull View v, Callback callback) {
        Objects.requireNonNull(v, "v can't be null");
        long lastClickTime = clickTimeArr.get(v.getId());
        if (System.currentTimeMillis() - lastClickTime >= minInterval) {
            callback.onAccept();
            clickTimeArr.put(v.getId(), System.currentTimeMillis());
        } else {
            callback.onReject();
        }
    }
    
    public interface Callback {
        /**
         * 拒绝
         */
        void onReject();

        /**
         * 通过
         */
        void onAccept();
    }
}

File: library/src/main/java/com/snail/widget/listener/RejectFastClickListener.kt
package com.snail.widget.listener

import android.util.SparseLongArray
import android.view.View

/**
 * 防止快速点击的
 *
 * date: 2019-06-16 12:29
 * author: zengfansheng
 */
abstract class RejectFastClickListener(private val minInterval: Long = 1000) {
    private val clickTimeArr = SparseLongArray()

    protected fun check(v: View?, cb: (accept: Boolean) -> Unit) {
        if (v != null) {
            val lastClickTime = clickTimeArr[v.id]
            if (System.currentTimeMillis() - lastClickTime >= minInterval) {
                cb.invoke(true)
                clickTimeArr.put(v.id, System.currentTimeMillis())
            } else {
                cb.invoke(false)
            }
        }
    }
}